codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c

```diff
@@ -0,0 +1,322 @@
+
+#define tmax 2000
+#define maxbm 30
+#define maxch 3
+#define maxrng 250
+
+int fnum=0;
+
+struct RadarParm *prm;
+struct FitData *fit;
+
+struct OptionData opt;
+
+
+int rst_opterr (char *txt) {
+  fprintf(stderr,"Option not recognized: %s\n",txt);
+  fprintf(stderr,"Please try: trim_fit --help\n");
+  return(-1);
+}
+
+
+// function to return the index in an array
+int get_index(int a, int b, int c, int d, int aSize, int bSize, int cSize) {
+    return (d * aSize * bSize * cSize) + (c * aSize * bSize) + (b * aSize) + a; 
+}
+
+// structure to store the qflgs. Can be dynamically resized
+typedef struct {
+  int *value;
+  size_t used;
+  size_t size;
+} qflgData;
+
+
+void free_parameters(struct RadarParm *prm, struct FitData *fit, FILE *fp, qflgData *q)
+{
+  if (prm != NULL) RadarParmFree(prm);
+  if (fit != NULL) FitFree(fit);
+  if (fp !=NULL) free(fp);
+  if (q->value !=NULL) {
+    free(q->value);
+    q->value=NULL;
+  }
+}
+
+
+
+int main (int argc,char *argv[]) {
+  
+  unsigned char vb=0;
+  unsigned char help=0;
+  unsigned char option=0;
+  unsigned char version=0;
+
+  OptionAdd(&opt,"-help",'x',&help);
+  OptionAdd(&opt,"-option",'x',&option);
+  OptionAdd(&opt,"-version",'x',&version);
+  OptionAdd(&opt,"vb",'x',&vb);
+
+  int arg;
+  arg=OptionProcess(1,argc,argv,&opt,rst_opterr);
+
+  if (arg==-1) {
+    exit(-1);
+  }
+
+  if (help==1) {
+    OptionPrintInfo(stdout,hlpstr);
+    exit(0);
+  }
+  if (option==1) {
+    OptionDump(stdout,&opt);
+    exit(0);
+  }
+
+  if (version==1) {
+    OptionVersion(stdout);
+    exit(0);
+  }
+  
+  FILE *fp=NULL;
+  if (arg==argc) fp=stdin;
+  else fp=fopen(argv[arg],"r");
+  if (fp==NULL) {
+    fprintf(stderr,"File not found.\n");
+    exit(-1);
+  }
+  
+  //set origin.command field
+  time_t ctime;
+  int c,n;
+  char command[128];
+  char tmstr[40];
+  command[0]=0;
+  n=0;
+  for (c=0;c<argc;c++) {
+    n+=strlen(argv[c])+1;
+    if (n>127) break;
+    if (c !=0) strcat(command," ");
+    strcat(command,argv[c]);
+  }
+  
+  
+  prm=RadarParmMake();
+  fit=FitMake();
+  if (FitFread(fp,prm,fit)==-1) {
+    fprintf(stderr,"Error reading file\n");
+    exit(-1);
+  }
+
+  // Initial memory allocation to store qflgs
+  qflgData *qflgs=malloc(sizeof(qflgData));
+  qflgs->value = malloc(tmax*maxrng*maxch*maxbm * sizeof(int));
+  qflgs->used = 0;
+  qflgs->size = tmax*maxrng*maxch*maxbm;
+  
+  
+  // Read the FITACF file to populate the qflg array
+  int index;
+  int maxindex=-1;
+  int rng,bm,ch;
+  int nrec[maxbm][maxch]; // counts number of records for each beam/channel
+  memset(nrec,0,sizeof(nrec));
+  
+  do {
+  
+    bm=prm->bmnum;
+    ch=prm->channel;
+    for (rng=0;rng<prm->nrang;rng++) {
+      index=get_index(bm,ch,rng,nrec[bm][ch],maxbm,maxch,maxrng);
+      if (maxindex < index) maxindex=index;
+      if (qflgs->size < maxindex) {
+        qflgs->size += tmax*maxrng*maxch*maxbm;
+        qflgs->value = realloc(qflgs->value, qflgs->size * sizeof(int));
```

Comment:
Check if the value is not 0 and errno was not set 
https://pubs.opengroup.org/onlinepubs/009695399/functions/realloc.html